db/tags/utils.cc

```diff
@@ -53,4 +55,32 @@ future<> update_tags(service::migration_manager& mm, schema_ptr schema, std::map
     });
 }
 
+future<> modify_tags(service::migration_manager& mm, sstring ks, sstring cf,
+                     std::function<void(std::map<sstring, sstring>&)> modify) {
+    co_await mm.container().invoke_on(0, [ks = std::move(ks), cf = std::move(cf), modify = std::move(modify)] (service::migration_manager& mm) -> future<> {
+        // FIXME: the following needs to be in a loop. If mm.announce() below
+        // fails, we need to retry the whole thing.
+        auto group0_guard = co_await mm.start_group0_operation();
+        // After getting the schema-modification lock, we need to read the
+        // table's *current* schema - it might have changed before we got
+        // the lock, by some concurrent modification. If the table is gone,
+        // this will throw no_such_column_family.
+        schema_ptr s = mm.get_storage_proxy().data_dictionary().find_schema(ks, cf);
+        const std::map<sstring, sstring>* tags_ptr = get_tags_of_table(s);
+        std::map<sstring, sstring> tags;
+        if (tags_ptr) {
+            // tags_ptr is a constant pointer to schema data. To allow func()
+            // to modify the tags, we must make a copy.
+            tags = *tags_ptr;
+        }
+        modify(tags);
+        schema_builder builder(s);
+        builder.add_extension(tags_extension::NAME, ::make_shared<tags_extension>(tags));
+
+        auto m = co_await mm.prepare_column_family_update_announcement(builder.build(), false, std::vector<view_ptr>(), group0_guard.write_timestamp());
```

Comment:
just a note as I don't know this code at all but I see almost everywhere timestamp is generated immediately after start_group0_operation()

Reply:
I'm not sure I understand your question, so tell me if I don't answer it:
The idea of this "group0_guard" is that the entire schema read-modify-write operation will be done as one atomic operation. The schema modification will have a timestamp - and it was indeed chosen above when group0_guard was created. It doesn't matter if a few milliseconds (or whatever) passed since this time.
So I think this code is correct. Did you see something different "almost everywhere"?

Reply:
it's ok, just it was usually called sooner, I asked just in case of some "edge" case (I was thinking maybe yielding could break it or something)